KeyError when using ansible module

I am trying to use ansible module but I have the following error:
```
Ansible = <ansible>

    def test_route_to_filter(Ansible):
>       assert Ansible('command', 'ip route get 8.8.8.8 mark {{ conn_mark }} | grep {{ inventory_hostname }} ')['rc'] == 0

test_igp_route.py:14: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/alkahan/.local/lib/python2.7/site-packages/testinfra/modules/ansible.py:46: in __call__
    module_name, module_args, check=check, **kwargs)
/home/alkahan/.local/lib/python2.7/site-packages/testinfra/backend/ansible.py:58: in run_ansible
    **kwargs)
/home/alkahan/.local/lib/python2.7/site-packages/testinfra/utils/ansible_runner.py:212: in run
    host, module_name, module_args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <testinfra.utils.ansible_runner.AnsibleRunnerV2 object at 0x7f0ad128ac90>, host = 'filter-1', module_name = 'command'
module_args = 'ip route get 8.8.8.8 mark {{ conn_mark }} | grep {{ inventory_hostname }} ', kwargs = {'check': True}
action = {'args': 'ip route get 8.8.8.8 mark {{ conn_mark }} | grep {{ inventory_hostname }} ', 'module': 'command'}, play = filter-1
tqm = <ansible.executor.task_queue_manager.TaskQueueManager object at 0x7f0ad12993d0>, options = <testinfra.utils.ansible_runner.Options object at 0x7f0ad1299a50>
callback = <testinfra.utils.ansible_runner.Callback object at 0x7f0ad12688d0>

    def run(self, host, module_name, module_args=None, **kwargs):
        action = {"module": module_name}
        if module_args is not None:
            if module_name in ("command", "shell"):
                # Workaround https://github.com/ansible/ansible/issues/13862
                module_args = module_args.replace("=", "\\=")
            action["args"] = module_args
        play = ansible.playbook.play.Play().load({
            "hosts": host,
            "gather_facts": "no",
            "tasks": [{
                "action": action,
            }],
        }, variable_manager=self.variable_manager, loader=self.loader)
        tqm = None
        options = Options(**kwargs)
        callback = Callback()
        try:
            tqm = ansible.executor.task_queue_manager.TaskQueueManager(
                inventory=self.inventory,
                variable_manager=self.variable_manager,
                loader=self.loader,
                options=options,
                passwords=None,
                stdout_callback=callback,
            )
            tqm.run(play)
        finally:
            if tqm is not None:
                tqm.cleanup()
    
        if host in callback.unreachable:
            raise RuntimeError(
                "host unreachable: {}".format(callback.unreachable[host]))
>       return callback.contacted[host]["result"]
E       KeyError: u'filter-1'

/home/alkahan/.local/lib/python2.7/site-packages/testinfra/utils/ansible_runner.py:192: KeyError
```

--- a/testinfra/backend/ansible.py
+++ b/testinfra/backend/ansible.py
@@ -15,6 +15,7 @@ from __future__ import unicode_literals
 from __future__ import absolute_import
 
 import logging
+import pprint
 
 from testinfra.backend import base
 import testinfra.utils.ansible_runner as ansible_runner
@@ -52,10 +53,15 @@ class AnsibleBackend(base.BaseBackend):
         return result
 
     def run_ansible(self, module_name, module_args=None, **kwargs):
-        return ansible_runner.run(
+        result = ansible_runner.run(
             self.host, module_name, module_args,
             host_list=self.ansible_inventory,
             **kwargs)
+        logger.info(
+            "RUN Ansible(%s, %s, %s): %s",
+            repr(module_name), repr(module_args), repr(kwargs),
+            pprint.pformat(result))
+        return result
 
     def get_variables(self):
         return ansible_runner.get_variables(

--- a/testinfra/modules/ansible.py
+++ b/testinfra/modules/ansible.py
@@ -14,8 +14,27 @@
 from __future__ import absolute_import
 from __future__ import unicode_literals
 
+import pprint
 
 from testinfra.modules.base import InstanceModule
+
+
+class AnsibleException(Exception):
+    """Exception raised when an error occur in an ansible call
+
+    result from ansible can be accessed through the ``result`` attribute
+
+    >>> try:
+    ...     Ansible("command", "echo foo")
+    ... except Ansible.AnsibleException as exc:
+    ...     assert exc.result['failed'] is True
+    ...     assert exc.result['msg'] == 'check mode not supported for command'
+    """
+
+    def __init__(self, result):
+        self.result = result
+        super(AnsibleException, self).__init__(
+            "Unexpected error: {}".format(pprint.pformat(result)))
 
 
 class Ansible(InstanceModule):
@@ -30,20 +49,26 @@ class Ansible(InstanceModule):
 
     >>> Ansible("apt", "name=nginx state=present")["changed"]
     False
+    >>> Ansible("command", "echo foo", check=False)["stdout"]
+    'foo'
     >>> Ansible("setup")["ansible_facts"]["ansible_lsb"]["codename"]
     'jessie'
     >>> Ansible("file", "path=/etc/passwd")["mode"]
     '0640'
 
     """
+    AnsibleException = AnsibleException
 
     def __call__(self, module_name, module_args=None, check=True, **kwargs):
         if not self._backend.HAS_RUN_ANSIBLE:
             raise RuntimeError((
                 "Ansible module is only available with ansible "
                 "connection backend"))
-        return self._backend.run_ansible(
+        result = self._backend.run_ansible(
             module_name, module_args, check=check, **kwargs)
+        if result.get("failed", False) is True:
+            raise AnsibleException(result)
+        return result
 
     def get_variables(self):
         """Returns a dict of ansible variables

--- a/testinfra/test/test_modules.py
+++ b/testinfra/test/test_modules.py
@@ -259,6 +259,8 @@ def test_ansible_unavailable(Ansible):
 
 @pytest.mark.testinfra_hosts("ansible://debian_jessie")
 def test_ansible_module(TestinfraBackend, Ansible):
+    import ansible
+    version = int(ansible.__version__.split(".", 1)[0])
     setup = Ansible("setup")["ansible_facts"]
     assert setup["ansible_lsb"]["codename"] == "jessie"
     passwd = Ansible("file", "path=/etc/passwd")
@@ -275,6 +277,29 @@ def test_ansible_module(TestinfraBackend, Ansible):
     variables = Ansible.get_variables()
     assert variables["inventory_hostname"] == "debian_jessie"
     assert variables["group_names"] == ["ungrouped"]
+
+    # test errors reporting
+    with pytest.raises(Ansible.AnsibleException) as excinfo:
+        Ansible("file", "path=/etc/passwd an_unexpected=variable")
+    tb = str(excinfo.value)
+    assert 'unsupported parameter for module: an_unexpected' in tb
+
+    with pytest.raises(Ansible.AnsibleException) as excinfo:
+        Ansible("command", "zzz")
+    if version == 1:
+        msg = "check mode not supported for command"
+    else:
+        msg = "Skipped. You might want to try check=False"
+    assert excinfo.value.result['msg'] == msg
+
+    try:
+        Ansible("command", "zzz", check=False)
+    except Ansible.AnsibleException as exc:
+        assert exc.result['rc'] == 2
+        assert exc.result['msg'] == '[Errno 2] No such file or directory'
+
+    result = Ansible("command", "echo foo", check=False)
+    assert result['stdout'] == 'foo'
 
 
 @pytest.mark.destructive

--- a/testinfra/utils/ansible_runner.py
+++ b/testinfra/utils/ansible_runner.py
@@ -16,6 +16,8 @@
 from __future__ import unicode_literals
 from __future__ import absolute_import
 
+import pprint
+
 try:
     import ansible
 except ImportError:
@@ -92,6 +94,9 @@ class AnsibleRunnerV1(AnsibleRunnerBase):
             **kwargs).run()
         if host not in result["contacted"]:
             raise RuntimeError("Unexpected error: {}".format(result))
+        if result["contacted"][host].get("skipped", False) is True:
+            # For consistency with ansible v2 backend
+            result["contacted"][host]["failed"] = True
         return result["contacted"][host]
 
 
@@ -113,24 +118,29 @@ class Options(object):
 if _has_ansible and _ansible_major_version == 2:
     class Callback(ansible.plugins.callback.CallbackBase):
 
-        def __init__(self):
-            self.unreachable = {}
-            self.contacted = {}
+        def __init__(self, *args, **kwargs):
+            self.result = {}
+            super(Callback, self).__init__(*args, **kwargs)
 
         def runner_on_ok(self, host, result):
-            self.contacted[host] = {
-                "success": True,
-                "result": result,
-            }
+            self.result = result
 
         def runner_on_failed(self, host, result, ignore_errors=False):
-            self.contacted[host] = {
-                "success": False,
-                "result": result,
-            }
+            self.result = result
 
+        # pylint: disable=no-self-use
         def runner_on_unreachable(self, host, result):
-            self.unreachable[host] = result
+            raise RuntimeError(
+                'Host {} is unreachable: {}'.format(
+                    host, pprint.pformat(result)),
+            )
+
+        def runner_on_skipped(self, host, item=None):
+            self.result = {
+                'failed': True,
+                'msg': 'Skipped. You might want to try check=False',
+                'item': item,
+            }
 
 
 class AnsibleRunnerV2(AnsibleRunnerBase):
@@ -186,10 +196,7 @@ class AnsibleRunnerV2(AnsibleRunnerBase):
             if tqm is not None:
                 tqm.cleanup()
 
-        if host in callback.unreachable:
-            raise RuntimeError(
-                "host unreachable: {}".format(callback.unreachable[host]))
-        return callback.contacted[host]["result"]
+        return callback.result
 
 if not _has_ansible:
     AnsibleRunner = AnsibleRunnerUnavailable

